packages/devtools_server/lib/src/server.dart

```diff
@@ -92,14 +97,82 @@ void serveDevTools({
 
   final server = await shelf.serve(handler, '127.0.0.1', port);
 
+  final devtoolsUrl = 'http://${server.address.host}:${server.port}';
+
   printOutput(
-    'Serving DevTools at http://${server.address.host}:${server.port}',
+    'Serving DevTools at $devtoolsUrl',
     {
       'method': 'server.started',
       'params': {'host': server.address.host, 'port': server.port, 'pid': pid}
     },
     machineMode: machineMode,
   );
+
+  _listenForUris(devtoolsUrl);
+}
+
+void _listenForUris(String devtoolsUrl) {
+  final Stream<Map<String, dynamic>> _stdinCommandStream = stdin
+      .transform<String>(utf8.decoder)
+      .transform<String>(const LineSplitter())
+      .where((String line) => line.startsWith('[{') && line.endsWith('}]'))
+      .map<Map<String, dynamic>>((String line) {
+    line = line.substring(1, line.length - 1);
+    return json.decode(line) as Map<String, dynamic>;
+  });
+
+  // Example json input: [{"url":"ws://localhost:8888/ws"}]
```

Comment:
I think this JSON should match the format of our other protocols and the output we send, with a `method` param, something like:

```js
{
  "method": "vm.register",
  "params": { "url": "blah://blah" }
}
```

Reply:
I wonder if we should also check `machineMode`? If we're not running in machine mode, maybe we don't need to parse this (or, we might want the interface to be simpler to paste a URL than to make it into JSON).

Reply:
I missed out the `ID` from the example above. To work the same as our other machine modes (which is essentially JSON-RPC), the dialog would look like this:

Editor sends the request (which includes an ID):

```js
{
  "id": 1,
  "method": "vm.register",
  "params": { "url": "blah://blah" }
}
```

Then the server should respond with either a result (which can be empty or null in this case, we don't really have any data to return):

```js
{
  "id": 1,
  "result": {}
}
```

or an error:

```js
{
  "id": 1,
  "error": "Unable to connect to VM service"
}
```

The code that handles this kind of thing for Flutter is [here](https://github.com/flutter/flutter/blob/d775908c7e95a5f00afa21fb92ce03ef65bc6563/packages/flutter_tools/lib/src/commands/daemon.dart). The error message should be formatted such that it's reasonable for the editor to show the user to tell them something has gone wrong.

Reply:
+1 on keeping the protocol consistent with other daemon style tools.
For now we only need to support the daemon version as this is for IDE users. Command line users are responsible for opening their own browser.

Reply:
changed this to expect json in the above format.